File: src/main/java/org/xi/quick/codegenerator/config/AppConfig.java
package org.xi.quick.codegenerator.config;

import freemarker.template.Template;
import freemarker.template.TemplateExceptionHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.xi.quick.codegenerator.model.FreemarkerModel;
import org.xi.quick.codegenerator.properties.FieldProperties;
import org.xi.quick.codegenerator.properties.FileProperties;
import org.xi.quick.codegenerator.properties.PathProperties;
import org.xi.quick.codegenerator.properties.GeneratorProperties;
import org.xi.quick.codegenerator.utils.DirectoryUtil;
import org.xi.quick.codegenerator.utils.SystemUtil;

import java.io.*;
import java.util.*;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Configuration
public class AppConfig {

    @Autowired
    GeneratorProperties generator;

    @Autowired
    PathProperties generatorPath;

    @Autowired
    FileProperties generatorFile;

    @Bean(name = "freeMarkerConfiguration")
    public freemarker.template.Configuration getConfiguration() throws IOException {

        File directory = new File(generatorPath.getTemplate());

        freemarker.template.Configuration cfg = new freemarker.template.Configuration(freemarker.template.Configuration.VERSION_2_3_24);
        cfg.setDirectoryForTemplateLoading(directory);
        cfg.setDefaultEncoding(generator.getEncoding());
        cfg.setTemplateExceptionHandler(TemplateExceptionHandler.RETHROW_HANDLER);

        return cfg;
    }

    /**
     * 所有模版
     *
     * @return
     * @throws IOException
     */
    @Bean(name = "allTemplates")
    public List<FreemarkerModel> getAllTemplates(freemarker.template.Configuration freeMarkerConfiguration) throws IOException {

        List<FreemarkerModel> templates = getMatchingTemplates(freeMarkerConfiguration,
                templateRelativePath ->
                        isClassFile(templateRelativePath)
                                && !isMatchingFile(templateRelativePath, generatorFile.getIgnore())
                                && !isMatchingFile(templateRelativePath, generatorFile.getAggregate())
                                && !isMatchingFile(templateRelativePath, generatorFile.getJustCopy()));
        return templates;
    }

    /**
     * 所有运行一次的模版
     *
     * @return
     * @throws IOException
     */
    @Bean(name = "allOnceTemplates")
    public List<FreemarkerModel> getAllOnceTemplates(freemarker.template.Configuration freeMarkerConfiguration) throws IOException {

        List<FreemarkerModel> templates = getMatchingTemplates(freeMarkerConfiguration,
                templateRelativePath ->
                        !isClassFile(templateRelativePath)
                                && !isMatchingFile(templateRelativePath, generatorFile.getIgnore())
                                && !isMatchingFile(templateRelativePath, generatorFile.getAggregate())
                                && !isMatchingFile(templateRelativePath, generatorFile.getJustCopy()));
        return templates;
    }

    /**
     * 所有聚合的模版
     *
     * @return
     * @throws IOException
     */
    @Bean(name = "allAggregateTemplates")
    public List<FreemarkerModel> getAllAggregateTemplates(freemarker.template.Configuration freeMarkerConfiguration) throws IOException {

        List<FreemarkerModel> templates = getMatchingTemplates(freeMarkerConfiguration,
                templateRelativePath ->
                        isMatchingFile(templateRelativePath, generatorFile.getAggregate()));
        return templates;
    }

    /**
     * 所有仅复制的文件
     *
     * @return
     * @throws IOException
     */
    @Bean(name = "allJustCopyTemplates")
    public List<FreemarkerModel> getAllJustCopyTemplates(freemarker.template.Configuration freeMarkerConfiguration) throws IOException {

        List<FreemarkerModel> templates = getMatchingTemplates(freeMarkerConfiguration,
                templateRelativePath ->
                        isMatchingFile(templateRelativePath, generatorFile.getJustCopy()));
        return templates;
    }


    /**
     * 获取匹配的模版
     *
     * @param freeMarkerConfiguration
     * @param predicate
     * @return
     * @throws IOException
     */
    public List<FreemarkerModel> getMatchingTemplates(freemarker.template.Configuration freeMarkerConfiguration,
                                                      Predicate<String> predicate) throws IOException {

        File directory = new File(generatorPath.getTemplate());
        String dirAbsolutePath = directory.getAbsolutePath();

        List<File> files = DirectoryUtil.getAllFiles(generatorPath.getTemplate());
        List<FreemarkerModel> result = new ArrayList<>();

        for (File file : files) {

            if (file.isHidden()) continue;

            String templateRelativePath = file.getAbsolutePath().substring(dirAbsolutePath.length() + 1);
            if (!predicate.test(templateRelativePath)) continue;

            Template template = freeMarkerConfiguration.getTemplate(templateRelativePath, generator.getEncoding());

            FreemarkerModel outModel = new FreemarkerModel(templateRelativePath, template);

            result.add(outModel);
        }

        return result;
    }

    /**
     * 是否是匹配模版
     *
     * @param templateRelativePath
     * @return
     */
    private boolean isMatchingFile(String templateRelativePath, Set<String> files) {

        for (String file : files) {
            if (templateRelativePath.startsWith(file) || templateRelativePath.contains(SystemUtil.SYSTEM_SLASH + file))
                return true;
        }

        return false;
    }

    /**
     * 是否是类相关文件
     *
     * @param path
     * @return
     */
    private static boolean isClassFile(String path) {

        Pattern pattern = Pattern.compile("\\$\\{className(_[^\\}]*)?\\}");
        Matcher matcher = pattern.matcher(path);
        return matcher.find();
    }

}

File: mysql-code-generator-kt/src/main/kotlin/org/xi/quick/codegeneratorkt/config/AppConfig.kt
package org.xi.quick.codegeneratorkt.config

import freemarker.template.TemplateExceptionHandler
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.xi.quick.codegeneratorkt.model.FreemarkerModel
import org.xi.quick.codegeneratorkt.properties.FileProperties
import org.xi.quick.codegeneratorkt.properties.GeneratorProperties
import org.xi.quick.codegeneratorkt.properties.PathProperties
import org.xi.quick.codegeneratorkt.utils.DirectoryUtils
import org.xi.quick.codegeneratorkt.utils.SystemUtils
import java.io.File
import java.io.IOException
import java.util.ArrayList

@Configuration
class AppConfig {

    @Autowired
    lateinit var generator: GeneratorProperties

    @Autowired
    lateinit var generatorPath: PathProperties

    @Autowired
    lateinit var generatorFile: FileProperties

    @Bean(name = ["freeMarkerConfiguration"])
    @Throws(IOException::class)
    fun getConfiguration(): freemarker.template.Configuration {
        val directory = File(generatorPath.template)

        val cfg = freemarker.template.Configuration(freemarker.template.Configuration.VERSION_2_3_28)
        cfg.setDirectoryForTemplateLoading(directory)
        cfg.defaultEncoding = generator.encoding
        cfg.templateExceptionHandler = TemplateExceptionHandler.RETHROW_HANDLER

        return cfg
    }

    /**
     * 所有模版
     *
     * @return
     * @throws IOException
     */
    @Bean(name = ["allTemplates"])
    @Throws(IOException::class)
    fun getAllTemplates(freeMarkerConfiguration: freemarker.template.Configuration): List<FreemarkerModel> {

        return getMatchingTemplates(freeMarkerConfiguration,
                { templateRelativePath ->
                    (isClassFile(templateRelativePath)
                            && !isMatchingFile(templateRelativePath, generatorFile.ignore)
                            && !isMatchingFile(templateRelativePath, generatorFile.aggregate)
                            && !isMatchingFile(templateRelativePath, generatorFile.justCopy))
                })
    }

    /**
     * 所有运行一次的模版
     *
     * @return
     * @throws IOException
     */
    @Bean(name = ["allOnceTemplates"])
    @Throws(IOException::class)
    fun getAllOnceTemplates(freeMarkerConfiguration: freemarker.template.Configuration): List<FreemarkerModel> {

        return getMatchingTemplates(freeMarkerConfiguration,
                { templateRelativePath ->
                    !isClassFile(templateRelativePath)
                            && !isMatchingFile(templateRelativePath, generatorFile.ignore)
                            && !isMatchingFile(templateRelativePath, generatorFile.aggregate)
                            && !isMatchingFile(templateRelativePath, generatorFile.justCopy)
                })
    }

    /**
     * 所有聚合的模版
     *
     * @return
     * @throws IOException
     */
    @Bean(name = ["allAggregateTemplates"])
    @Throws(IOException::class)
    fun getAllAggregateTemplates(freeMarkerConfiguration: freemarker.template.Configuration): List<FreemarkerModel> {

        return getMatchingTemplates(freeMarkerConfiguration,
                { templateRelativePath -> isMatchingFile(templateRelativePath, generatorFile.aggregate) })
    }

    /**
     * 所有仅复制的文件
     *
     * @return
     * @throws IOException
     */
    @Bean(name = ["allJustCopyTemplates"])
    @Throws(IOException::class)
    fun getAllJustCopyTemplates(freeMarkerConfiguration: freemarker.template.Configuration): List<FreemarkerModel> {

        return getMatchingTemplates(freeMarkerConfiguration,
                { templateRelativePath -> isMatchingFile(templateRelativePath, generatorFile.justCopy) })
    }


    /**
     * 获取匹配的模版
     *
     * @param freeMarkerConfiguration
     * @param predicate
     * @return
     * @throws IOException
     */
    @Throws(IOException::class)
    fun getMatchingTemplates(freeMarkerConfiguration: freemarker.template.Configuration,
                             predicate: (String) -> Boolean): List<FreemarkerModel> {

        val directory = File(generatorPath.template)
        val dirAbsolutePath = directory.absolutePath

        val files = DirectoryUtils.getAllFiles(generatorPath.template)
        val result = ArrayList<FreemarkerModel>()

        for (file in files) {

            if (file.isHidden) continue

            val templateRelativePath = file.absolutePath.substring(dirAbsolutePath.length + 1)
            if (!predicate(templateRelativePath)) continue

            val template = freeMarkerConfiguration.getTemplate(templateRelativePath, generator.encoding)

            val outModel = FreemarkerModel(templateRelativePath, template)

            result.add(outModel)
        }

        return result
    }

    /**
     * 是否是匹配模版
     *
     * @param templateRelativePath
     * @return
     */
    private fun isMatchingFile(templateRelativePath: String, files: Set<String>?): Boolean {

        if (files != null) {
            for (file in files) {
                if (templateRelativePath.startsWith(file) || templateRelativePath.contains(SystemUtils.SYSTEM_SLASH + file))
                    return true
            }
        }

        return false
    }

    /**
     * 是否是类相关文件
     *
     * @param path
     * @return
     */
    private fun isClassFile(path: String): Boolean {
        return Regex("""\$\{className(_[^\}]*)?\}""").containsMatchIn(path)
    }

}